File: backend/src/main/java/com/cephalononni/service/AuthService.java
package com.cephalononni.service;

import com.cephalononni.exception.ApiException;
import com.cephalononni.model.User;
import com.cephalononni.model.UserRole;
import com.cephalononni.repository.UserRepository;
import com.cephalononni.security.JwtService;
import com.cephalononni.web.dto.AuthDtos.LoginRequest;
import com.cephalononni.web.dto.AuthDtos.RegisterRequest;
import com.cephalononni.web.dto.AuthDtos.UserPublic;
import org.springframework.lang.NonNull;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;

    public AuthService(UserRepository userRepository, PasswordEncoder passwordEncoder, JwtService jwtService) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtService = jwtService;
    }

    /**
     * Creates a new Tenno user. Email is only checked for `@` and `.` presence (a loose
     * substring check, not real format validation), and must not already be registered.
     */
    @Transactional
    public UserPublic register(RegisterRequest request) {
        String email = request.email().trim();
        if (!email.contains("@") || !email.contains(".")) {
            throw ApiException.badRequest("Invalid email format");
        }
        if (userRepository.existsByEmail(email)) {
            throw ApiException.badRequest("Email already used");
        }

        User user = new User();
        user.setEmail(email);
        user.setUsername(request.username().trim());
        user.setHashedPassword(passwordEncoder.encode(request.password()));
        user.setRole(UserRole.TENNO);
        user = userRepository.save(user);

        return toPublic(user);
    }

    public record LoginResult(UserPublic user, @NonNull String token) {
    }

    /** Verifies email/password and issues a JWT for the matching user, or throws 401. */
    public LoginResult login(LoginRequest request) {
        String email = request.email() == null ? "" : request.email();
        String password = request.password() == null ? "" : request.password();

        User user = userRepository.findByEmail(email).orElse(null);
        if (user == null || !passwordEncoder.matches(password, user.getHashedPassword())) {
            throw ApiException.unauthorized("Invalid credentials");
        }
        String token = jwtService.issueToken(user.getId());
        return new LoginResult(toPublic(user), token);
    }

    /** Maps a user entity to its wire representation. */
    public static UserPublic toPublic(User user) {
        return new UserPublic(String.valueOf(user.getId()), user.getEmail(), user.getUsername(), user.getRole().getWireValue());
    }
}

File: backend-python-legacy/app/services/auth_service.py
from typing import Optional

from database.dynamic.auth import create_token
from database.dynamic.security import hash_password, verify_password
from models.users import UserRole
from repositories.user_repository import UserRepository


class AuthService:
    @staticmethod
    async def register(email: str, username: str, password: str) -> dict:
        existing = await UserRepository.find_by_email(email)
        if existing:
            raise ValueError("Email already used")

        user_data = {
            "email": email,
            "username": username,
            "hashed_password": hash_password(password),
            "role": UserRole.TENNO.value,
        }

        user = await UserRepository.create(user_data)
        return {
            "id": str(user["_id"]),
            "email": user["email"],
            "username": user["username"],
            "role": user["role"],
        }

    @staticmethod
    async def login(email: str, password: str) -> dict:
        user = await UserRepository.find_by_email(email)
        if not user:
            raise ValueError("Invalid credentials")

        if not password or not verify_password(password, user["hashed_password"]):
            raise ValueError("Invalid credentials")

        token = create_token({"sub": str(user["_id"])})
        return {
            "token": token,
            "id": str(user["_id"]),
            "email": user["email"],
            "username": user["username"],
            "role": user.get("role", UserRole.TENNO.value),
        }

    @staticmethod
    async def get_current_user(user_id: str) -> Optional[dict]:
        user = await UserRepository.find_by_id(user_id)
        if not user:
            return None
        return {
            "id": str(user["_id"]),
            "email": user["email"],
            "username": user["username"],
            "role": user.get("role", UserRole.TENNO.value),
        }
